### PR TITLE
Added a rule of thumb for when to use invariant, contravariant, and cova...

### PR DESCRIPTION
**Pull request by jeremieSimon**
Added a rule of thumb for when to use invariant, contravariant, and covariant type. 
I can also add an extra explanation if needed about contravariant as it is missing right now.

### CODE REVIEW

File: effectivescala.mo [hunk 1]
```diff
@@ -301,6 +301,10 @@ is typically invalid with mutable collections. Consider
   *	show contravariance trick?
 -->
 
+More generally, types should be made covariant when you only get or produce an element.
```
Comment (mosesn) [inline effectivescala.mo]: These seem to describe when it's *possible* to make a types covariant / contravariant, but I'm not sure it provides a clear rationale for why you should do it whenever possible.

Comment (jeremieSimon) [inline effectivescala.mo]: That's true. Let me submit another PR and be more specific on that matter.

**Comment by jeremieSimon** [comment]
@mosesn , I think it should be now more clear. 
Let me know what you think. 

Thanks

**Comment by mosesn** [comment]
The example isn't quite right.

For A, B, C, D, where A <: B and C <: D,
B => C is a A => D, but A => C is not a B => D.

NB: Your Translator class should not compile.

I'm also not sure that this really provides guidance on best practices in scala–it seems more like it explains how to operate the type system.  Is it appropriate for Effective Scala?

**Comment by jeremieSimon** [comment]
Yes I realized that afterwards :)
And true, it might not be so appropriate for effective scala.
I just wanted to fill a blank that there is in the current guideline right now about contravariance.

**Comment by mosesn** [comment]
:+1: Yeah, it's a good idea.  Maybe we can revisit in the future.
